src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -440,18 +440,10 @@ void Lowering::LowerBlockStore(GenTreeBlk* blkNode)
 
         if (blkNode->OperIs(GT_STORE_OBJ))
         {
-            if (!blkNode->AsObj()->GetLayout()->HasGCPtr())
+            if (!blkNode->AsObj()->GetLayout()->HasGCPtr() || (isDstAddrLocal && (size <= copyBlockUnrollLimit)))
             {
                 blkNode->SetOper(GT_STORE_BLK);
             }
-            else if (isDstAddrLocal && (size <= copyBlockUnrollLimit))
```

Comment:
It seems like we don't need the `isDstAddrLocal` check: why can't we use this code path even if copying with src=heap, dst=heap, or src=stack, dst=heap?

Reply:
I was trying to consolidate the `blkNode->SetOper(GT_STORE_BLK)` case, but in general, we need to disable GC for following `unroll` cases:
- isDstAddrLocal = true (as of today)
- isDstAddrLocal = false (if dst is on heap)
- isSrcAddrLocal = false (if src is on heap)
- isSrcAddrLocal = true (since we want to disable GC regardless if dst is on stack/heap, src state won't matter).

So basically, if we are doing `GT_STORE_BLK` and unrolling, we need to disable GC. With that said, I think we should disable the GC inside `genCodeForCpBlkUnroll()` only for cases where we will be doing the adjustments i.e. `srcOffsetAdjustment != 0 || dstOffsetAdjustment != 0` to limit it to scenarios that are really affected.

Reply:
With that being the case, I missed a case for objects that has GC pointers. Fixed in https://github.com/dotnet/runtime/pull/70053